File: backend/app/api/v1/routes/posts.py
from fastapi import APIRouter, Depends, HTTPException
from app.core.dependencies import get_current_user
from app.modules.posts.service import PostService
from app.modules.posts.dependencies import get_post_service
from app.modules.posts.schemas import (
    CreatePostRequest,
    AddCommentRequest,
    PostResponse,
    CommentResponse,
    UploadUrlResponse,
)

router = APIRouter()


ALLOWED_IMAGE_TYPES = {"image/jpeg", "image/png", "image/webp", "image/gif", "image/heic"}

@router.get("/upload-url", response_model=UploadUrlResponse)
def get_upload_url(
    filename: str,
    content_type: str = "image/jpeg",
    current_user_id: str = Depends(get_current_user),
    service: PostService = Depends(get_post_service),
):
    if content_type not in ALLOWED_IMAGE_TYPES:
        raise HTTPException(status_code=400, detail=f"Unsupported file type: {content_type}")
    return service.get_upload_url(current_user_id, filename, content_type)


@router.post("/", response_model=PostResponse, status_code=201)
def create_post(
    body: CreatePostRequest,
    current_user_id: str = Depends(get_current_user),
    service: PostService = Depends(get_post_service),
):
    return service.create(current_user_id, body)


@router.get("/{post_id}", response_model=PostResponse)
def get_post(
    post_id: str,
    service: PostService = Depends(get_post_service),
):
    return service.get_post(post_id)


@router.delete("/{post_id}", status_code=204)
def delete_post(
    post_id: str,
    current_user_id: str = Depends(get_current_user),
    service: PostService = Depends(get_post_service),
):
    service.delete(post_id, current_user_id)


@router.post("/{post_id}/like", status_code=204)
def like_post(
    post_id: str,
    current_user_id: str = Depends(get_current_user),
    service: PostService = Depends(get_post_service),
):
    service.like(post_id, current_user_id)


@router.delete("/{post_id}/like", status_code=204)
def unlike_post(
    post_id: str,
    current_user_id: str = Depends(get_current_user),
    service: PostService = Depends(get_post_service),
):
    service.unlike(post_id, current_user_id)


@router.post("/{post_id}/comments", response_model=CommentResponse, status_code=201)
def add_comment(
    post_id: str,
    body: AddCommentRequest,
    current_user_id: str = Depends(get_current_user),
    service: PostService = Depends(get_post_service),
):
    return service.comment(post_id, current_user_id, body.text)


@router.get("/{post_id}/comments", response_model=list[CommentResponse])
def get_comments(
    post_id: str,
    service: PostService = Depends(get_post_service),
):
    return service.get_post_comments(post_id)

File: frontend/src/lib/posts.ts
import { api } from "./api";

export interface Post {
  post_id: string;
  user_id: string;
  caption: string;
  image_url: string;
  likes_count: number;
  created_at: string;
}

export interface Comment {
  comment_id: string;
  post_id: string;
  user_id: string;
  text: string;
  created_at: string;
}

export async function getUploadUrl(filename: string): Promise<{ upload_url: string; image_url: string }> {
  const { data } = await api.get("/v1/posts/upload-url", { params: { filename } });
  return data;
}

export async function uploadToS3(uploadUrl: string, file: File): Promise<void> {
  await fetch(uploadUrl, {
    method: "PUT",
    body: file,
    headers: { "Content-Type": file.type },
  });
}

export async function createPost(caption: string, image_url: string): Promise<Post> {
  const { data } = await api.post("/v1/posts/", { caption, image_url });
  return data;
}

export async function deletePost(postId: string): Promise<void> {
  await api.delete(`/v1/posts/${postId}`);
}

export async function likePost(postId: string): Promise<void> {
  await api.post(`/v1/posts/${postId}/like`);
}

export async function unlikePost(postId: string): Promise<void> {
  await api.delete(`/v1/posts/${postId}/like`);
}

export async function addComment(postId: string, text: string): Promise<Comment> {
  const { data } = await api.post(`/v1/posts/${postId}/comments`, { text });
  return data;
}

export async function getComments(postId: string): Promise<Comment[]> {
  const { data } = await api.get(`/v1/posts/${postId}/comments`);
  return data;
}
